workflows_core/engine/dense_output_engine.py

```diff
@@ -0,0 +1,140 @@
+"""
+    Stable Engine Pseudo-algorithm-
+        1. Downloads about 1000 documents.
+        2. Processes it in much smaller chunks.
+        3. Upserts all 1000 documents.
+        4. Repeat until dataset has finished looping
+
+    We download a large chunk and upsert large chunks to avoid hammering
+    our servers.
+
+"""
+import logging
+import traceback
+
+from typing import Optional, List, Sequence
+
+from workflows_core.dataset.dataset import Dataset
+from workflows_core.operator.abstract_operator import AbstractOperator
+from workflows_core.engine.abstract_engine import AbstractEngine
+from workflows_core.utils.document import Document
+from workflows_core.types import Filter
+
+from tqdm.auto import tqdm
+
+logger = logging.getLogger(__file__)
+
+
+class DenseOutputEngine(AbstractEngine):
+    def __init__(
+        self,
```

Comment:
Can you add a function that stores the relationship between the output datasets and the input dataset